### PR TITLE
Add On-demand WAL Download to logicalfuncs

### PR DESCRIPTION
**Pull request by save-buffer**
We implemented on-demand WAL download for walsender, but other things that may want to read the WAL from safekeepers don't do that yet. This PR makes it do that by adding the same set of hooks to logicalfuncs. 

Addresses https://github.com/neondatabase/neon/issues/7959

Also relies on:
https://github.com/neondatabase/postgres/pull/438
https://github.com/neondatabase/postgres/pull/437
https://github.com/neondatabase/postgres/pull/436

### CODE REVIEW

**Comment by jcsp** [comment]
I'm a reviewer on this PR, but I have no idea what it's about.  Please can you write a description that fills out the PR template -- the other PR you linked has an empty description.

**Comment by save-buffer** [comment]
Sorry, I did have an issue that links to a slack thread but I didn't paste correctly. I've added a description

**Comment by kelvich** [comment]
@save-buffer that PR would benefit from a test that can reproduce the issue without a fix and would pass with fix :-)

**Comment by kelvich** [comment]
Error message looks like that:

```
PG:2024-06-06 16:22:46.537 GMT ttid=1d880e086b24349586dc1ee10ebea34e/1921f4934124c56e8e937fefffe88bfa
sqlstate=XX000 [6370]
ERROR:  invalid magic number 0000 in log segment 00000001000000000000000F, offset 12206080
```

(posting here so that search on the error message could find this issue; originally i went in https://github.com/neondatabase/cloud/issues/12992).

**Comment by kelvich** [comment]
Also double-checked that user faces `invalid magic number 0000 in log segment` when calling `pg_logical_slot_peek_binary_changes`, which is turn calls `pg_logical_slot_get_changes_guts`. So everything checks out. https://neonprod.grafana.net/goto/D-SdG_sIR?orgId=1

**Comment by save-buffer** [comment]
Added a test

**Comment by save-buffer** [comment]
Without callbacks: 
```
psycopg2.errors.InternalError_: invalid record length at 0/2ADA620: wanted 24, got 0
```

With callbacks: test passes

**Review by kelvich — APPROVED** [review]

**Comment by kelvich** [comment]
@knizhnik and/or @ars pls take a look too as you are marked as code owners here

File: test_runner/regress/test_logical_replication.py
Comment (arssher) [inline test_runner/regress/test_logical_replication.py:242]: I don't see what is the point of generating a segment of WAL data if slot is created later; pg_logical_slot_peek_binary_changes won't stream these 16MB anyway, so for this to make sense slot should be created before bulk insertion.

But in principle just restart of compute is enough to test on demand download as current segment on the new compute would be empty.

**Review by arssher — APPROVED** [review]
Thanks! Let's also run pgindent on C part.

File: pgxn/neon/neon.c
Comment (save-buffer) [reply pgxn/neon/neon.c]: Moved, need approval on 
https://github.com/neondatabase/postgres/pull/444
https://github.com/neondatabase/postgres/pull/442
https://github.com/neondatabase/postgres/pull/443

File: test_runner/regress/test_logical_replication.py
Comment (save-buffer) [reply test_runner/regress/test_logical_replication.py]: Good point, I moved it above. Test still seems to pass

File: pgxn/neon/neon.c
Comment (save-buffer) [reply pgxn/neon/neon.c]: (CI will be broken until those merge and I update revisions.json)

Comment (save-buffer) [reply pgxn/neon/neon.c]: Should be updated now

**Review by sharnoff — APPROVED** [review]

**Pull request merged**
